chain/client/src/archive/cold_store_actor.rs

```diff
@@ -0,0 +1,496 @@
+    /// If migration fails sleeps for 30s and tries again.
+    /// If migration returned any successful status (including interruption status) breaks the loop.
+    fn cold_store_migration_loop(&self, ctx: &mut dyn DelayedActionRunner<Self>) {
+        if !self.keep_going.load(std::sync::atomic::Ordering::Relaxed) {
+            tracing::debug!(target: "cold_store", "stopping the initial migration loop");
+            return;
+        }
+
+        match self.cold_store_migration() {
+            Err(err) => {
+                // We can either stop the cold store thread or hope that next time migration will not fail.
+                // Here we pick the second option.
+                let dur =
+                    self.split_storage_config.cold_store_initial_migration_loop_sleep_duration;
+                tracing::error!(target: "cold_store", ?err, ?dur, "Migration failed. Sleeping and trying again.");
+                ctx.run_later("cold_store_migration_loop", dur, move |actor, ctx| {
+                    actor.cold_store_migration_loop(ctx);
+                });
+            }
+            Ok(migration_status) => {
+                match migration_status {
+                    ColdStoreMigrationResult::MigrationInterrupted => {
+                        tracing::info!(target: "cold_store", "Cold storage migration was interrupted");
+                        return;
+                    }
+                    ColdStoreMigrationResult::NoNeedForMigration
+                    | ColdStoreMigrationResult::SuccessfulMigration => {
+                        // Migration was successful, we can start the cold store loop.
+                        tracing::info!(target : "cold_store", "Starting the cold store loop");
+                        self.cold_store_loop(ctx);
+                    }
+                }
+            }
+        }
+    }
+
+    // This method will copy data from hot storage to cold storage in a loop.
+    // It will try to copy blocks as fast as possible up until cold head = final head.
+    // Once the cold head reaches the final head it will sleep for one second before
+    // trying to copy data at the next height.
+    fn cold_store_loop(&self, ctx: &mut dyn DelayedActionRunner<Self>) {
+        if !self.keep_going.load(std::sync::atomic::Ordering::Relaxed) {
+            tracing::debug!(target : "cold_store", "Stopping the cold store loop");
+            return;
+        }
+
+        let result = self.cold_store_loop_impl();
+
+        // A block older than the final head was copied. We should continue copying
+        // until cold head reaches final head.
+        let dur = if let Ok(ColdStoreCopyResult::OtherBlockCopied) = result {
+            Duration::ZERO
+        } else {
+            self.split_storage_config.cold_store_loop_sleep_duration
+        };
+
+        ctx.run_later("cold_store_loop", dur, move |actor, ctx| {
+            actor.cold_store_loop(ctx);
+        });
```

Comment:
Run later with duration zero seems a bit odd but I guess it's fine?

Reply:
It's helpful as if the actor is stopped the next call isn't scheduled